src/Sources/AbstractSource.php

```diff
@@ -135,44 +161,100 @@ function ($migration_class) {
         $this->onUnInstall();
     }
 
-    protected function save(Model $me, Model $primary)
+    /**
+     * Checks for new primary entity and source info, updated source info, and updates the audit table as needed.
+     *
+     * @param Model $source
+     * @param $primary
+     * @param $existing_source
+     * @param array $sourceColumns
+     * @return void
+     */
+    protected function save(Model $source, $primary, $existing_source, array $sourceColumns)
     {
-        // @todo: update $me name
-        // @todo: Only save model if something has changed
-        // @todo: in the Audit, describe what changed
-        // Now, we save the primary version
-        $primary->save();
+        // Create new primary entity and source if game doesn't already exist.
+        if ($primary !== null) {
+            $primary->save();
+            $column_name = static::getPrimaryEntityJoinColumnName();
+            $source->{$column_name} = $primary->id;
+            $source->save();
+        }
 
-        //@todo: update game_id to more generic
-        $me->game_id = $primary->id;
+        // Update source if changes are found.
+        if ($existing_source !== null) {
+            $changes = $this->findChanges($source, $existing_source, $sourceColumns);
+            $source->id = $existing_source->id;
+            $existing_source->save();
+        } else {
+            $changes = null;
+        }
 
-        // Create new record in this source
-        $me->save();
+        // Creates a new Audit Record for new source info or if something changed.
+        $this->updateAuditTable($source, $changes);
+    }
+
+    /**
+     * Checks for and only saves new and updated data.
+     *
+     * @param $source
+     * @param $changes
+     * @return false|void
+     */
+    public function updateAuditTable($source, $changes)
+    {
+        if ($changes === []) {
+            return false;
+        }
 
-        // Create a new Audit Record because something changed
         $audit = new Audit();
         $audit->job_id = $this->getJob()->id;
-        $audit->record_id = $me->id;
+        $audit->record_id = $source->id;
         $audit->source_key = static::getKey();
 
+        if ($changes !== null) {
```

Comment:
What may work better here is to put all of this inside the loop -- everything from 209 to 230. And then, for every change, a unique audit record is created.

That also means you don't need to check if changes is `null`. You can just set it to an empty array up earlier and then loop through the empty array here -- doing nothing.

Reply:
That also means you don't need to combine the values into a string like you are doing.